File: src/morphic/frontend/routes_converter.py
"""
Converter-tab API routes — scan, formats, convert, delete, progress.
"""

from __future__ import annotations

import os
import threading
import time
import uuid

from flask import Blueprint, jsonify, request

from morphic.converter.constants import IMAGE_CONVERSIONS, VIDEO_CONVERSIONS
from morphic.converter.converter import convert_file
from morphic.converter.scanner import scan_folder
from morphic.shared.utils import format_file_size

bp = Blueprint("converter", __name__)

# ── In-memory progress store ───────────────────────────────────────────────

_conversion_jobs: dict[str, dict] = {}
_jobs_lock = threading.Lock()


# ── Scan ───────────────────────────────────────────────────────────────────


@bp.route("/scan", methods=["POST"])
def api_scan():
    """Scan a folder for convertible media files."""
    data = request.get_json()
    if not data:
        return jsonify({"error": "No JSON body"}), 400

    folder = data.get("folder", "").strip()
    if not folder or not os.path.isdir(folder):
        return jsonify({"error": f"Invalid folder: {folder}"}), 400

    include_sub = data.get("include_subfolders", True)
    filter_type = data.get("filter_type", "both")
    if filter_type not in ("images", "videos", "both"):
        filter_type = "both"

    result = scan_folder(
        folder,
        include_subfolders=include_sub,
        filter_type=filter_type,
    )
    return jsonify(result)


# ── Formats ────────────────────────────────────────────────────────────────


@bp.route("/formats")
def api_formats():
    """Return all conversion mappings (source -> targets)."""
    return jsonify({
        "image": IMAGE_CONVERSIONS,
        "video": VIDEO_CONVERSIONS,
    })


# ── Convert ────────────────────────────────────────────────────────────────


@bp.route("/convert", methods=["POST"])
def api_convert():
    """Convert files and return progress job id."""
    data = request.get_json()
    if not data:
        return jsonify({"error": "No JSON body"}), 400

    files = data.get("files", [])
    target_ext = data.get("target_ext", "").strip()
    delete_original = data.get("delete_original", False)

    if not files or not target_ext:
        return jsonify({"error": "files and target_ext required"}), 400

    job_id = str(uuid.uuid4())[:8]
    job = {
        "id": job_id,
        "status": "running",
        "total": len(files),
        "completed": 0,
        "current_file": "",
        "results": [],
        "error": None,
    }
    with _jobs_lock:
        _conversion_jobs[job_id] = job

    thread = threading.Thread(
        target=_run_conversion,
        args=(job, files, target_ext, delete_original),
        daemon=True,
    )
    thread.start()
    return jsonify({"job_id": job_id}), 202


def _run_conversion(
    job: dict,
    files: list[str],
    target_ext: str,
    delete_original: bool,
) -> None:
    """Background worker for batch conversion."""
    for i, source in enumerate(files):
        job["current_file"] = os.path.basename(source)
        try:
            original_size = os.path.getsize(source) if os.path.isfile(source) else 0
            dest = convert_file(source, target_ext)
            new_size = os.path.getsize(dest) if os.path.isfile(dest) else 0

            result = {
                "source": source,
                "destination": dest,
                "status": "ok",
                "original_size": original_size,
                "new_size": new_size,
                "original_size_fmt": format_file_size(original_size),
                "new_size_fmt": format_file_size(new_size),
            }

            if delete_original and os.path.isfile(source):
                try:
                    os.remove(source)
                    result["source_deleted"] = True
                except OSError:
                    result["source_deleted"] = False
        except Exception as e:
            result = {
                "source": source,
                "destination": None,
                "status": "error",
                "error": str(e),
            }
        job["results"].append(result)
        job["completed"] = i + 1

    job["status"] = "done"
    job["current_file"] = ""


# ── Progress ───────────────────────────────────────────────────────────────


@bp.route("/progress/<job_id>")
def api_progress(job_id: str):
    """Get progress of a conversion job."""
    with _jobs_lock:
        job = _conversion_jobs.get(job_id)
    if not job:
        return jsonify({"error": "Job not found"}), 404
    return jsonify(job)


@bp.route("/progress/<job_id>/poll")
def api_progress_poll(job_id: str):
    """Long-poll for conversion progress (blocks up to 10s)."""
    with _jobs_lock:
        job = _conversion_jobs.get(job_id)
    if not job:
        return jsonify({"error": "Job not found"}), 404

    # Lightweight polling — wait until progress changes or timeout
    last_completed = int(request.args.get("last", -1))
    deadline = time.time() + 10
    while time.time() < deadline:
        if job["completed"] != last_completed or job["status"] == "done":
            return jsonify(job)
        time.sleep(0.3)
    return jsonify(job)


# ── Delete (standalone) ───────────────────────────────────────────────────


@bp.route("/delete", methods=["POST"])
def api_delete():
    """Delete selected files."""
    data = request.get_json()
    if not data:
        return jsonify({"error": "No JSON body"}), 400

    files = data.get("files", [])
    if not files:
        return jsonify({"error": "No files specified"}), 400

    results = []
    total_freed = 0

    for fp in files:
        try:
            if not os.path.isfile(fp):
                results.append({"path": fp, "status": "not_found"})
                continue
            size = os.path.getsize(fp)
            os.remove(fp)
            total_freed += size
            results.append({
                "path": fp,
                "status": "deleted",
                "size_freed": size,
            })
        except PermissionError:
            results.append({"path": fp, "status": "permission_denied"})
        except Exception as e:
            results.append({"path": fp, "status": "error", "error": str(e)})

    return jsonify({
        "results": results,
        "total_freed": total_freed,
        "total_freed_formatted": format_file_size(total_freed),
    })

File: web/routes_converter.go
package web

import (
	"net/http"
	"os"
	"time"

	"github.com/gin-gonic/gin"
	"github.com/exterex/morphic/internal/converter"
	"github.com/exterex/morphic/internal/shared"
)

var conversionStore = shared.NewJobStore[conversionJob]()

type conversionJob struct {
	shared.Job
	Total       int                      `json:"total"`
	Completed   int                      `json:"completed"`
	CurrentFile string                   `json:"current_file"`
	Results     []map[string]interface{} `json:"results"`
}

func init() {
	conversionStore.StartCleanup(30*time.Minute, func(j *conversionJob) time.Time {
		return j.DoneAt
	})
}

func registerConverterRoutes(r *gin.Engine) {
	g := r.Group("/api/converter")
	{
		g.POST("/scan", handleConverterScan)
		g.GET("/formats", handleConverterFormats)
		g.POST("/convert", handleConverterConvert)
		g.GET("/progress/:id", handleConverterProgress)
		g.GET("/progress/:id/poll", handleConverterPoll)
		g.POST("/delete", handleConverterDelete)
	}
}

func handleConverterScan(c *gin.Context) {
	var req struct {
		Folder            string `json:"folder"`
		IncludeSubfolders *bool  `json:"include_subfolders"`
		FilterType        string `json:"filter_type"`
	}
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}
	if req.Folder == "" || !isDir(req.Folder) {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid folder: " + req.Folder})
		return
	}
	includeSub := true
	if req.IncludeSubfolders != nil {
		includeSub = *req.IncludeSubfolders
	}
	filterType := req.FilterType
	if filterType != "images" && filterType != "videos" && filterType != "both" {
		filterType = "both"
	}

	result, err := converter.ScanFolder(req.Folder, includeSub, filterType)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}
	c.JSON(http.StatusOK, result)
}

func handleConverterFormats(c *gin.Context) {
	c.JSON(http.StatusOK, gin.H{
		"image": converter.ImageConversions,
		"video": converter.VideoConversions,
	})
}

func handleConverterConvert(c *gin.Context) {
	var req struct {
		Files          []string `json:"files"`
		TargetExt      string   `json:"target_ext"`
		DeleteOriginal bool     `json:"delete_original"`
		AV1CRF         *int     `json:"av1_crf"`
	}
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}
	if len(req.Files) == 0 || req.TargetExt == "" {
		c.JSON(http.StatusBadRequest, gin.H{"error": "files and target_ext required"})
		return
	}

	av1CRF := 0
	if req.AV1CRF != nil {
		av1CRF = *req.AV1CRF
	}

	job := &conversionJob{
		Job:   shared.NewJob(),
		Total: len(req.Files),
	}
	job.Status = shared.JobStatusRunning
	conversionStore.Set(job.ID, job)

	go runConversion(job, req.Files, req.TargetExt, req.DeleteOriginal, av1CRF)

	c.JSON(http.StatusAccepted, gin.H{"job_id": job.ID})
}

func runConversion(job *conversionJob, files []string, targetExt string, deleteOriginal bool, av1CRF int) {
	for i, source := range files {
		job.CurrentFile = source

		result := map[string]interface{}{
			"source":         source,
			"source_deleted": false,
		}

		origSize := int64(0)
		if info, err := os.Stat(source); err == nil {
			origSize = info.Size()
		}

		dest, err := converter.ConvertFile(source, targetExt, "", av1CRF)
		if err != nil {
			result["destination"] = nil
			result["status"] = "error"
			result["error"] = err.Error()
		} else {
			newSize := int64(0)
			if info, err := os.Stat(dest); err == nil {
				newSize = info.Size()
			}

			result["destination"] = dest
			result["status"] = "ok"
			result["original_size"] = origSize
			result["new_size"] = newSize
			result["original_size_fmt"] = shared.FormatFileSize(origSize)
			result["new_size_fmt"] = shared.FormatFileSize(newSize)

			// Delete original only if explicitly requested and safe
			if deleteOriginal && dest != "" {
				absSrc, _ := absPath(source)
				absDest, _ := absPath(dest)
				if absSrc != absDest && newSize > 0 {
					if err := os.Remove(source); err == nil {
						result["source_deleted"] = true
					}
				}
			}
		}

		job.Results = append(job.Results, result)
		job.Completed = i + 1
	}

	job.Status = shared.JobStatusDone
	job.CurrentFile = ""
	job.DoneAt = time.Now()
}

func handleConverterProgress(c *gin.Context) {
	id := c.Param("id")
	job, ok := conversionStore.Get(id)
	if !ok {
		c.JSON(http.StatusNotFound, gin.H{"error": "Job not found"})
		return
	}
	c.JSON(http.StatusOK, gin.H{
		"id":           job.ID,
		"status":       job.Status,
		"total":        job.Total,
		"completed":    job.Completed,
		"current_file": job.CurrentFile,
		"results":      job.Results,
		"error":        job.Error,
	})
}

func handleConverterPoll(c *gin.Context) {
	id := c.Param("id")
	job, ok := conversionStore.Get(id)
	if !ok {
		c.JSON(http.StatusNotFound, gin.H{"error": "Job not found"})
		return
	}

	lastStr := c.Query("last")
	last := -1
	if lastStr != "" {
		for i := 0; i < len(lastStr); i++ {
			if lastStr[i] >= '0' && lastStr[i] <= '9' {
				last = last*10 + int(lastStr[i]-'0')
			}
		}
	}

	deadline := time.Now().Add(10 * time.Second)
	for time.Now().Before(deadline) {
		if job.Completed != last || job.Status == shared.JobStatusDone {
			break
		}
		time.Sleep(300 * time.Millisecond)
	}

	c.JSON(http.StatusOK, gin.H{
		"id":           job.ID,
		"status":       job.Status,
		"total":        job.Total,
		"completed":    job.Completed,
		"current_file": job.CurrentFile,
		"results":      job.Results,
		"error":        job.Error,
	})
}

func handleConverterDelete(c *gin.Context) {
	var req struct {
		Files []string `json:"files"`
	}
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}
	if len(req.Files) == 0 {
		c.JSON(http.StatusBadRequest, gin.H{"error": "No files specified"})
		return
	}

	var results []map[string]interface{}
	totalFreed := int64(0)

	for _, fp := range req.Files {
		info, err := os.Stat(fp)
		if err != nil {
			results = append(results, map[string]interface{}{"path": fp, "status": "not_found"})
			continue
		}
		if info.IsDir() {
			results = append(results, map[string]interface{}{"path": fp, "status": "not_found"})
			continue
		}
		size := info.Size()
		if err := os.Remove(fp); err != nil {
			if os.IsPermission(err) {
				results = append(results, map[string]interface{}{"path": fp, "status": "permission_denied"})
			} else {
				results = append(results, map[string]interface{}{"path": fp, "status": "error", "error": err.Error()})
			}
		} else {
			totalFreed += size
			results = append(results, map[string]interface{}{"path": fp, "status": "deleted", "size_freed": size})
		}
	}

	c.JSON(http.StatusOK, gin.H{
		"results":              results,
		"total_freed":          totalFreed,
		"total_freed_formatted": shared.FormatFileSize(totalFreed),
	})
}

func absPath(p string) (string, error) {
	abs, err := os.Getwd()
	if err != nil {
		return p, err
	}
	if len(p) > 0 && p[0] == '/' {
		return p, nil
	}
	return abs + "/" + p, nil
}
